path_checkouts.go

```diff
@@ -0,0 +1,390 @@
+package openldap
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/helper/locksutil"
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+const checkoutKeyType = "checkout-creds"
+
+func (b *backend) pathSetCheckOut() []*framework.Path {
+	return []*framework.Path{
+		{
+			Pattern: libraryPrefix + framework.GenericNameRegex("name") + "/check-out$",
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeLowerCaseString,
+					Description: "Name of the set",
+					Required:    true,
+				},
+				"ttl": {
+					Type:        framework.TypeDurationSecond,
+					Description: "The length of time before the check-out will expire, in seconds.",
+				},
+			},
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.UpdateOperation: &framework.PathOperation{
+					Callback: b.operationSetCheckOut,
+					Summary:  "Check a service account out from the library.",
+				},
+			},
+			HelpSynopsis: `Check a service account out from the library.`,
+		},
+	}
+}
+
+func (b *backend) operationSetCheckOut(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	setName := fieldData.Get("name").(string)
+
+	lock := locksutil.LockForKey(b.checkOutLocks, setName)
+	lock.Lock()
+	defer lock.Unlock()
+
+	ttlPeriodRaw, ttlPeriodSent := fieldData.GetOk("ttl")
+	if !ttlPeriodSent {
+		ttlPeriodRaw = 0
+	}
+	requestedTTL := time.Duration(ttlPeriodRaw.(int)) * time.Second
+
+	set, err := readSet(ctx, req.Storage, setName)
+	if err != nil {
+		return nil, err
+	}
+	if set == nil {
+		return logical.ErrorResponse(fmt.Sprintf(`%q doesn't exist`, setName)), nil
+	}
+
+	// Prepare the check-out we'd like to execute.
+	ttl := set.TTL
+	if ttlPeriodSent {
+		switch {
+		case set.TTL <= 0 && requestedTTL > 0:
+			// The set's TTL is infinite and the caller requested a finite TTL.
+			ttl = requestedTTL
+		case set.TTL > 0 && requestedTTL < set.TTL:
+			// The set's TTL isn't infinite and the caller requested a shorter TTL.
+			ttl = requestedTTL
+		}
+	}
+	newCheckOut := &CheckOut{
+		IsAvailable:         false,
+		BorrowerEntityID:    req.EntityID,
+		BorrowerClientToken: req.ClientToken,
+	}
+
+	// Check out the first service account available.
+	for _, serviceAccountName := range set.ServiceAccountNames {
+		if err := b.CheckOut(ctx, req.Storage, serviceAccountName, newCheckOut); err != nil {
+			if err == errCheckedOut {
+				continue
+			}
+			return nil, err
+		}
```

Comment:
I noticed this doesn't also continue the loop if err is `errNotFound`. I'm thinking this might be because there shouldn't be a case where a name in the Library Set contains a service account name that isn't managed by the plugin. Is that logic correct or on the right track?

Reply:
Yep, that's correct! It shouldn't be possible to hit that case. If it happens, there would be a bug in the code somewhere.